### ISSUE TITLE
A type of VBL that has buttons to enable or disable it -- like a door or window.

### ISSUE THREAD

**Issue by adventuremagic123** [RPTools/maptool#2003]
**Is your feature request related to a problem? Please describe.**
I would like to make it easier to manage VBL when doors and windows are opened or closed.  I'd rather have this support native to MapTool rather than having to add something like BoT (bag of tricks), which comes with more stuff than I want.

**Describe the solution you'd like**
I would like to have a new type of VBL that I can draw on my maps that I can select and enable or disable by using buttons specifically for that purpose -- like for a door or window.  So that I can clearly distinguish this new type of VBL from the traditional VBL, this new VBL type should be drawn with a different color or maybe dashed lines.  The buttons to enable or disable it would be on the tool bar.  A button might be provided for this new type of VBL -- to give me access to drawing tools similar to the existing VBL and access to the new enable and disable buttons.  Design this feature however you want.  I'm really just trying to convey an idea.

**Describe alternatives you've considered**
I currently delete and redraw VBL as needed.  It would be nice to not have to do that.  I could probably use Bag of Tricks to do this, but I'd rather have the support native in MapTool.  Bag of Tricks comes with a lot of other things that I don't really want to add to my campaigns.

**Additional context**
Foundary VTT has something that appears to be like this -- I'd use toolbar buttons, though, to enable and disable it rather than have buttons on the map.  You can see it in the YouTube video.

https://foundryvtt.com/

https://www.youtube.com/watch?v=7fu48nvvUY4&feature=youtu.be

**Comment by Azhrei** [comment]
Have you tried the automatic doors that Reverend has posted in the #artwork channel on Discord?  The VBL is attached to the door, so macros to open/close the doors automatically move the VBL too.  They're pretty cool. :)

**Comment by adventuremagic123** [comment]
I'm not finding it, although I did find as much as a few people talking about it.

My preference would be to have something like what I've described built into MapTool.  It's really the behavior I'm looking for.  I think #2004 describes something not yet supported by anything, and I could use that if this feature weren't implemented.

**Comment by bubblobill** [comment]
[Smart_Doors_v1.3.3.zip for MT v1.7+](https://github.com/bubblobill/MapTool/blob/master/SmartDoors/Smart_Doors_v1.3.3.zip)
[Smart_Doors_v1.3.3_compatible.zip for MT <v1.7](https://github.com/bubblobill/MapTool/blob/master/SmartDoors/Smart_Doors_v1.3.3_compatible.zip)

**Comment by Azhrei** [comment]
At some point soon-ish, these doors will be added to the art packs for MT or perhaps just made part of the default resources.

**Comment by adventuremagic123** [comment]
Thanks for the info.  I'll look at it.  I think part of the issue with these other solutions is that they feel like too much of a work-around.  It's been a while since I've looked at it.  I'll review the solution above and BoT and provide feedback.

**Comment by adventuremagic123** [comment]
I appreciate all the work that's gone into BoT and SmartDoors, but I don't want to have to put specialized tokens on my maps to get these kinds of features.  My maps already have doors and windows.  I don't want the extra work using these kinds of things create.  Instead, barring any native solution in MapTool, I'd rather just erase and redraw VBL and MBL.  I'm hoping #2003 and especially #2004 get enough traction to be implemented.  I'd really love to do it that way.

**Comment by adventuremagic123** [comment]
I think there was some confusion with what I was requesting.  I think the way MapTool currently does VBL and MBL is great -- and never should be changed!  It's different than, for example, what Foundry VTT has -- and that is a good thing!  This is the MapTool way!

What I'm saying is that when we markup our maps, VBL lays down one type of blocking layer (blue) and MBL lays down another (red).  Use both, and right now you get purple.

What I want is to introduce another type of BL (blocking layer) where we can use all the currently existing blocking layer tools to lay it down and erase it.  However!  I'm asking for a new behavior for this type of blocking layer.  I want to be able to use the standard blocking layer tools to -- while I'm running a game -- enable or disable this type of blocking layer.  This layer may actually exist as two types:  transparent and opaque.  There might be a better choice of terminology, but these might be called a "window blocking layer" and a "door blocking layer" -- so WBL and DBL.

When drawing WBL, the "W" would be used instead of "M" or "V" or "V/M".  Similarly, "D" would be used for DBL.

WBL might be an opaque yellow, and DBL might be an opaque orange -- when enabled!  When disabled, these might be made 50% opaque (so the GM can tell where they are)!

Most commonly, I'd probably use the blocking layer filled-in-square tool to enable or disable the WBL and DBL.

So, the problem is that now we have new modes of manipulating VBL.

**THEREFORE, HERE'S AN EVEN BETTER IDEA:**

VBL can be used as DBL, and MBL can be used as WBL.  How?  Well, if we had a **new control** that indicates when we use the **blocking layer tools whether we are:

1. Erasing
2. Drawing
3. Enabling
4. Disabling**

**We could then disable and enable selected VBL and MBL (or both).**  When **enabled**, VBL and MBL show a **100% opaque color**.  When **disabled**, VBL and MBL show a **50% opaque color -- and when it that state are ignored by the blocking layer**.

Here's an illustration of VBL and MBL set to the "enabled" state, so they are displayed as 100% opaque colors and are heeded by the blocking layer (so vision is blocked for the VBL and movement is blocked for the MBL).

![example](https://user-images.githubusercontent.com/50708405/89347012-994e9780-d66f-11ea-9700-1ef5def29595.JPG)

But, consider what could happen if during a game I switched back to the blocking layer tools, somehow indicated that the mode I wanted was to "disable", and used the filled-in-square-block tool to disable a region of VBL/MBL over one of my doors -- so the tokens could see and move into a room!

With everything we can currently do to edit blocking layer, it seems like the current MapTool implementation could support this concept as it is currently designed.  Please consider!

**Comment by JamzTheMan** [comment]
That's a lot to digest lol...

BUT, my gut is thinking you are saying "As a user, I want to be able to more easily and quickly enable/disable mbl/vbl"

I'm not sure we're up for adding more "layers" as that's just going to confuse end users even more?  But I do get the request. I think we still want to hold off until a later release with a renderer rewrite.  Adding "layers" means more java awt area intersections which is already pretty expensive to do (cpu intense).

I'd like to move us to using individual polygons as we move towards ray casting, along with drag-able/editable points and a slightly better interface to draw/show/hide/view all of the BL's...

**Comment by Phergus** [comment]
I _think_ the key point is that he doesn't want to use tokens with VBL.  Instead he wants to draw *BL and then be able to turn it off and back on _without having to redraw it._   So we would need to be able to turn on/off individual segments of a VBL path/polygon.

**Comment by JamzTheMan** [comment]
> I _think_ the key point is that he doesn't want to use tokens with VBL. Instead he wants to draw *BL and then be able to turn it off and back on _without having to redraw it._ So we would need to be able to turn on/off individual segments of a VBL path/polygon.

Yep, thats my sense too. And we could do that better by having polygons on the BL layers vs one combined one and add UI elements/interactions with said polygons.

The JWT lib has a multipolygon class and IF we were to do this under the current code base I would move us to that and that would gain us speed and other improvements as well BUT not sure its worth doing atm if we are going to redesign MT...

**Comment by adventuremagic123** [comment]
If you want to or see advantages to using polygons, that's fine.

Though I like the way MapTool currently works.  Just like you can use blocking layer drawing tools to erase VBL and MBL, I'd like to be able to use them to disable and enable VBL and MBL.  I'm NOT asking to be able to enable or disable specific segments like Foundry VTT does.  If you really feel it's the right direction, that's fine.

**Comment by JamzTheMan** [comment]
ya, it sounds like you want to be able to turn on/off the whole layer. but add layers so you can turn off just one of them for windows/doors. but doesn't make sense that ALL the doors/windows would turn on/off at the same time.

But I get that the "token vbl" isn't as easy as it could be.

**Comment by Phergus** [comment]
@adventuremagic123 

>  I'm NOT asking to be able to enable or disable specific segments like Foundry VTT does.

Actually that is what you asked for.

> I would like to have a new type of VBL and MBL that I can draw on my maps **that I can select and enable or disable** by using buttons specifically for that purpose

> Most commonly, **I'd probably use the blocking layer filled-in-square tool to enable or disable** the WBL and DBL.

If you were to use the rectangle tool to highlight an arbitrary section of the WBL/DBL, that is selecting one or more specific segments and possibly creating new ones.

Anyways, some more user-friendly way to handle VBL for doors and windows when they are embedded in the map image instead of as separate objects is needed.  What that is will need to be determined.
